### ISSUE TITLE
EPERM, operation not permitted error on Win7 after upgrade to 0.6.5...

### ISSUE THREAD

**Issue by DavidCRivera** [sdarnell/meteor-win#19]
Overall, Meteor is running fine, and it's great to finally have a solution that will keep us more up to date. But, since switching to the latest bootstraper from the MSI, I've been getting intermittent crashes that appear to have something to do with the bundling/building of my app. Here is what I'm getting in cmd.exe run as Administrator:

```
=> Meteor server restarted (x14)
C:\Users\username\AppData\Local\.meteor\tools\36f9aeca0e\tools\run.js:42
5
    throw e;
          ^
Error: EPERM, operation not permitted 'c:\path\file.htm
l~RFd60ee2.TMP'
    at Object.fs.statSync (fs.js:524:18)
    at readDirectory (C:\Users\username\AppData\Local\.meteor\tools\36f9
aeca0e\tools\watch.js:229:22)
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\username\AppData\Local\.meteo
r\tools\36f9aeca0e\lib\node_modules\underscore\underscore.js:79:11)
    at Object.readDirectory (C:\Users\username\AppData\Local\.meteor\too
ls\36f9aeca0e\tools\watch.js:224:5)
    at _.extend._fireIfDirectoryChanged (C:\Users\username\AppData\Local
\.meteor\tools\36f9aeca0e\tools\watch.js:343:31)
    at FSWatcher._.extend._startDirectoryWatches (C:\Users\username\AppD
ata\Local\.meteor\tools\36f9aeca0e\tools\watch.js:439:18)
    at FSWatcher.EventEmitter.emit (events.js:99:17)
    at FSEvent._handle.onchange (fs.js:826:12)

```

If I restart Meteor, it works fine with no errors. Maybe something related to temp files and directory permissions? I really don't know. I tried googling, but couldn't find this exact error.

Thanks in advance for the help.

UPDATE:

Probably related, I get this error in the server console although the site continues to run without any noticeable problems:

```
W20130909-13:03:04.136(-5)? (STDERR) Error: EBADF, read
```

**Comment by sdarnell** [comment]
First of all, the way file watching occurs in Meteor core changed drastically (completely re-written) in 0.6.5, and it is now more precise, whereas previously it was pretty much just looking for any old change. So I can understand if the upshot is that it works differently in some environments.

The first thing it is worth considering is what is generating those temporary files, as I'm pretty sure it isn't Meteor. My guess is that it is your text editor as it either saves changes to a file (using an atomic rename way of updating the file). However, it may be just a temporary file that it keeps for some other reason.
What editor do you use?
Can you configure it to put temporary files elsewhere?

The next issue is regardless of what is creating those files, why is node fs.statSync() giving EPERM?
My guess is that it is not because you don't have access rights, but more "operation not permitted" (at this time).
I have seen antivirus products sometimes lock files while they scan them, preventing an apparently valid operation.
But I wouldn't expect that sort of error from a stat operation (which is essentially just a read).

Internally node (uv library) does a CreateFile(FILE_READ_ATTRIBUTES) followed by GetFileInformationByHandle().
My reading of the win32 error mapping code suggests that the two win32 errors that produce EPERM are: ERROR_ACCESS_DENIED and ERROR_PRIVILEGE_NOT_HELD. We can probably ignore ERROR_PRIVILEGE_NOT_HELD.

The MSDN page for CreateFile ( http://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx ) has a specific mention that I think may be relevant:

> If you call CreateFile on a file that is pending deletion as a result of a previous call to DeleteFile, the function fails. The operating system delays file deletion until all handles to the file are closed. GetLastError returns ERROR_ACCESS_DENIED.

Given this is a temporary file it seems quite reasonable that this is happening.

So I think the solution is probably to make Meteor's watcher more resilient to EPERM. That bit of code already has a work-around for another error code (ENOENT). So adding an EPERM clause should be OK.

One other point worth noting is that the file watcher already excludes/ignores certain files (see bundler.js line 176). Including files ending in ~ (tilde). Which is a common backup file suffix for editors (vim, emacs, nano). So one solution may be to hack builder.js to add an extra regexp `/\.tmp$/` to the ignore list. Unfortunately I'm not sure it is common enough to add by default, and I can't see any way of getting developer specific ignores added (seems reasonable to me to add some developer or project specific ignore patterns.

For reference, the two win32 errors that produce EBADF are: ERROR_INVALID_HANDLE ERROR_INVALID_FLAGS, though I think we can ignore ERROR_INVALID_FLAGS.

**Comment by DavidCRivera** [comment]
Thanks for the detailed explanation. I'm using Sublime Text 3. I'll look into settings involving temp files and get back to you.

UPDATE: This page seems to cover this issue: http://www.sublimetext.com/forum/viewtopic.php?f=3&t=11610
I'm going to try that.

UPDATE: That was it. Over several hours of development it hasn't crashed with a similar error once. So the answer (from the above link) is to turn off the 'atomic save' setting in Sublime Text. 

I do still get the 

```
W20130910-13:02:51.330(-5)? (STDERR) Error: EBADF, read
```

error, but no noticeable effects like before.

Once again, thanks for the in depth answer. You got me where I needed to be to fix this. Really appreciate all the hard work!

**Comment by sdarnell** [comment]
I have submitted a change that ignore EPERM from statSync() in 7f20e79a8d37b1f8b4558c433feb88258eae889b
This will be released in 0.6.6.
